src/analyserexternalvariable.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "libcellml/analyserexternalvariable.h"
+
+#include "libcellml/component.h"
+
+#include "analyserexternalvariable_p.h"
+#include "utilities.h"
+
+namespace libcellml {
+
+AnalyserExternalVariable::AnalyserExternalVariableImpl::AnalyserExternalVariableImpl(const VariablePtr &variable)
+    : mVariable(variable)
+{
+}
+
+std::vector<VariableWeakPtr>::iterator AnalyserExternalVariable::AnalyserExternalVariableImpl::findDependency(const ModelPtr &model,
+                                                                                                              const std::string &componentName,
+                                                                                                              const std::string &variableName)
+{
+    return std::find_if(mDependencies.begin(), mDependencies.end(), [=](const VariableWeakPtr &v) {
+        auto variable = v.lock();
+
+        return (variable != nullptr)
+               && (owningModel(variable) == model)
+               && (owningComponent(variable)->name() == componentName)
+               && (variable->name() == variableName);
+    });
+}
+
+std::vector<VariableWeakPtr>::iterator AnalyserExternalVariable::AnalyserExternalVariableImpl::findDependency(const VariablePtr &variable)
+{
+    return std::find_if(mDependencies.begin(), mDependencies.end(), [=](const VariableWeakPtr &v) {
+        return v.lock() == variable;
+    });
+}
+
+AnalyserExternalVariable::AnalyserExternalVariable(const VariablePtr &variable)
+    : mPimpl(new AnalyserExternalVariableImpl(variable))
+{
+}
+
+AnalyserExternalVariable::~AnalyserExternalVariable()
+{
+    delete mPimpl;
+}
+
+AnalyserExternalVariablePtr AnalyserExternalVariable::create(const VariablePtr &variable) noexcept
+{
+    return std::shared_ptr<AnalyserExternalVariable> {new AnalyserExternalVariable {variable}};
+}
+
+VariablePtr AnalyserExternalVariable::variable() const
+{
+    return mPimpl->mVariable.lock();
+}
+
+bool AnalyserExternalVariable::addDependency(const VariablePtr &variable)
+{
+    auto pimplVariable = AnalyserExternalVariable::variable();
+
+    if ((pimplVariable != nullptr)
```

Comment:
Two things:
 - `AnalyserExternalVariable::variable()` is not a constructor. It's defined just above as:
```
VariablePtr AnalyserExternalVariable::variable() const
{
    return mPimpl->mVariable.lock();
}
```
   So, here, we are protecting ourselves against the case where `mVariable` is not available anymore.
 - We have 100% coverage, but we have an `A && B && C && D` test, so `return false` only occurs when `A`, `B`, `C` or `D` is false. Coverage is clearly not smart enough to cover all possible cases.As it happens, if we had the following code before that test:
```
if (pimplVariable == nullptr) {
    std::cout << ">>> pimplVariable is null!" << std::endl;
}
```
   We never get that message printed out... since we don't have a test that removes a variable before adding it (or one of its equivalent variables) as an external variable. I thought I would add such a test:
```
TEST(AnalyserExternalVariable, addRemovedVariable)
{
    auto parser = libcellml::Parser::create();
    auto model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));

    EXPECT_EQ(size_t(0), parser->issueCount());

    auto component = model->component("membrane");
    auto variable = component->variable("V");
    auto externalVariable = libcellml::AnalyserExternalVariable::create(variable);
    auto dependency = component->variable("Cm");

    component->removeVariable(variable);

    externalVariable->addDependency(dependency);

    EXPECT_EQ(variable, externalVariable->variable());
}
```
   but the message still doesn't get printed out. Indeed,although `variable` is removed, it's still being used twice somewhere else in libCellML...!? So, it looks like we should be using a weak `Variable` pointer in some places...

So, in summary, I feel like leaving things as they stand, just to be on the safe side.

Reply:
@agarny I left a comment somewhere asking about testing for constructor success ... this is the place I meant.  Since it's not a constructor (my bad) that other comment can be disregarded, but I still don't understand how we have coverage for something that you can't ... well ... cover?

Reply:
As far as I can tell, it's a limitation of the coverage tools we are using. Remember, we have an `if` statement that goes something like:
```
if (A && B && C && D) {
    // Do something...
    return true;
}
return false;
```
with `A` being `(pimplVariable != nullptr)`. To properly test the case where `A` is `false`, we would need to rewrite the `if` statement to read something like:
```
if (A) {
    if (B) {
        if (C) {
            if (D) {
                // Do something...
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    } else {
        return false;
    }
} else {
    return false;
}
```
I don't think we want to go that route, not least because I tried to test the case where `A` is `false` and it failed due to our not using weak pointers in all possible places.

Reply:
Hmmm ... maybe our coverage tools are different.  Mine gives me an error if I don't have all parts genuinely covered.  If I had your first statement, and `A` was always false, I'd get a coverage error.

Reply:
It's not 100% clear to me what you have tried, so can't really comment.